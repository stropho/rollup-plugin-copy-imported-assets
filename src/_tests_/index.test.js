
import babel from 'rollup-plugin-babel';
import { rollup } from 'rollup';
import copyImportedAssets from '../index';

const BABEL_OPTIONS = {
  babelrc: false,
  presets: [['@babel/preset-env', { modules: false }]],
  exclude: ['node_modules/**'],
};

const OUTPUT_OPTIONS = {
  file: 'output.js',
  format: 'esm',
};

describe('warnings', () => {
  it('warns about missing include/exclude options', async () => {
    const onwarnSpy = jest.fn();
    const inputOptions = {
      onwarn: onwarnSpy,
      input: 'src/_tests_/fixture/single-file/index.js',
      plugins: [
        copyImportedAssets(), // missing options !
        babel(BABEL_OPTIONS),
      ],
    };

    const bundle = await rollup(inputOptions);
    const { output } = await bundle.generate(OUTPUT_OPTIONS);

    expect(output.length).toBe(1);
    expect(onwarnSpy).toHaveBeenCalledTimes(1);
    expect(onwarnSpy.mock.calls[0][0].code).toBe('PLUGIN_WARNING');
    expect(onwarnSpy.mock.calls[0][0].message).toMatchSnapshot();
  });
  it('warns about unsupported output format', async () => {
    const onwarnSpy = jest.fn();
    const inputOptions = {
      onwarn: onwarnSpy,
      input: 'src/_tests_/fixture/single-file/index.js',
      plugins: [
        copyImportedAssets({ include: /.*/ }),
        babel(BABEL_OPTIONS),
      ],
    };

    const bundle = await rollup(inputOptions);
    const { output } = await bundle.generate({ ...OUTPUT_OPTIONS, format: 'commonjs' });

    expect(output.length).toBe(1);
    expect(onwarnSpy).toHaveBeenCalledTimes(1);
    expect(onwarnSpy.mock.calls[0][0].code).toBe('PLUGIN_WARNING');
    expect(onwarnSpy.mock.calls[0][0].message).toMatchSnapshot();
  });
});
describe('importing-asset', () => {
  it('generates esm output with external asset', async () => {
    const inputOptions = {
      input: 'src/_tests_/fixture/importing-asset/index.js',
      plugins: [
        copyImportedAssets({ include: /\.myext/ }),
        babel(BABEL_OPTIONS),
      ],
    };

    const bundle = await rollup(inputOptions);
    const { output } = await bundle.generate(OUTPUT_OPTIONS);

    expect(output.length).toBe(2);
    expect(output[0].imports.length).toBe(1);
    expect(output[0].code.includes('from \'./assets/asset-b7235ae7.myext\'')).toBeTruthy();

    expect(output[1].isAsset).toBeTruthy();
    expect(output[1].fileName.endsWith('/asset-b7235ae7.myext')).toBeTruthy();
  });
});

// describe('importing-asset-and-transforming', () => {
//   it('generates esm output with external css asset generated from other file', async () => {
//     const inputOptions = {
//       input: 'src/_tests_/fixture/importing-asset/index.js',
//       plugins: [
//         // {
//         //   resolveId(id, importer) {
//         //     // does not seem possible to transform files and send to next plugin
//         //   },
//         //   transform(code, id) {
//         //     // does not seem possible to transform files and send to next plugin
//         //   },
//         // },
//         copyImportedAssets({ include: /\.myext$/ }),
//         babel(BABEL_OPTIONS),
//       ],
//     };

//     const bundle = await rollup(inputOptions);
//     const { output } = await bundle.generate(OUTPUT_OPTIONS);

//     expect(output.length).toBe(2);
//     expect(output[0].imports.length).toBe(1);

//     expect(output[0].code.includes('from \'./assets/asset-b7235ae7.css\'')).toBeTruthy();
//   });
// });
