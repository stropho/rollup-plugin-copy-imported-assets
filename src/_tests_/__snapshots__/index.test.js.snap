// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing-asset generates esm output with external asset 1`] = `"assets/asset-f1618825.myext"`;

exports[`importing-asset-to-a-shared-chunk generates 4 files importing asset also in entry points 1`] = `"assets/asset-f1618825.myext"`;

exports[`importing-asset-to-a-shared-chunk generates 4 files importing asset only once 1`] = `"assets/asset-f1618825.myext"`;

exports[`importing-parent-asset generates esm output with external asset that has relative path 1`] = `"assets/asset-f1618825.myext"`;

exports[`importing-same-asset generates esm output with single external asset imported multiple times 1`] = `"assets/asset-f1618825.myext"`;

exports[`warnings warns about missing include/exclude options 1`] = `"No options to include files were specified - ignoring all"`;

exports[`warnings warns about unsupported output format 1`] = `"Only format 'esm' is supported, others should NOT be used with this plugin"`;
